### ISSUE TITLE
Trust-regions returns a point outside the manifold

### ISSUE THREAD

**Issue by IWalds** [JuliaManifolds/Manopt.jl#99]
Hello,

I have the following code, which defines an optimization problem on the oblique manifold, and calls the trust-regions solver on it. I expect the output to belong to the oblique manifold, but it is actually does not (as can be seen from computing the maximal column norm of the output matrix, which should be 1, but is around 1.4).

```julia
using Printf, Manifolds, ManifoldsBase, Manopt

n = 20
p = 2
M = Oblique(p,n)

# Cost matrix
C = zeros(n,n)
for k=1:n-1
    C[k,k+1] = 1 ;
    C[k+1,k] = 1 ;
end

Vinit = [ones(1,8) zeros(1,3) zeros(1,n-11) ;
         zeros(1,8) ones(1,3) -ones(1,n-11)] ;

function cost(M,V)
    # V -> Tr(C*V'*V)
    sum(vec((V*C) .* V))
end

function riem_grad(M,V)
    project(M,V,2*V*C')
end

function riem_hess(M,V,W)
    m = sum((V*C').*V,dims=1)
    WS = W * C - W .* m
    2. * project(M,V,WS)
end

@printf("Maximal column norm (before Trust Regions) : %.4f\n",
        maximum(sqrt.(sum(Vinit.^2,dims=1))))

Vrec = trust_regions(M,cost,riem_grad,riem_hess,Vinit;
                        stopping_criterion=StopAfterIteration(20))

@printf("Maximal column norm (after Trust Regions) : %.4f\n",
        maximum(sqrt.(sum(Vrec.^2,dims=1))))
```

I am unsure what the cause of the problem is. My impression is that there are numerical instabilities in the truncated conjugate gradient descent solver when the iterates come close to a degenerate critical point, which result in `β` becoming quite big, and in `o.δ` leaving the tangent space of the manifold.

If I add a projection step after the update of `o.δ` in the truncated conjugate gradient descent solver, the issue disappears, but I do not know if it is a sensible fix.
```diff
    o.residual = o.residual + α * o.Hδ
    # Precondition the residual.
    o.z = o.randomize ? o.residual : get_preconditioner(p, o.x, o.residual)
    zr = inner(p.M, o.x, o.z, o.residual)
    # Compute new search direction.
    β = zr / o.z_r
    o.z_r = zr
    o.δ = -o.z + β * o.δ
+    o.δ = project!(p.M,o.δ,o.x,o.δ)
```

**Comment by kellertuer** [comment]
HI,
thanks for posting the issue and great that you are using Manopt! Great that you directly posed an (n)MWE, though I will only have time to check the next few days.

Since the projection step seems to resolve the issue, you are probably right that we have a numerical instability here. (you could leave out the `o.δ =`part, since `project!` works in place of `o.δ` the way you wrote it).

The reason the projection is not there currently is twofold
1. the projection of tangent vectors might not be implemented on some manifold, so not requiring a projection makes the method available on more manifolds.
2. even if it is available, using it every iteration introduces more computational effort and might make the overall runtime longer.

I will try to find time the next few days and think about this. Maybe we can introduce an option to activate the projection for numerical stability. Let me now what you think.

PS: Is the last line the check whether you are on Oblique? You can also use `is_point(M,Vrec)` for this (or `is_point(M, Vrecm true)` in case you want to see the error if it is not on M.

**Comment by IWalds** [comment]
Hi,

Thank you very much for the reply. In my opinion, an option to activate a fix would be great. But as to the fix, the one I proposed in my initial message does not always work. Indeed, I found another setting this afternoon (the corresponding code is below) where the Trust Regions iterates leave the manifold, and where projecting `o.δ` onto the tangent space does not solve the issue. My impression in this case (but I did not have much time to look carefully at it) is that it is not about `β` being big; it seems that there is a numerical drift, maybe due to the fact that the small errors in the projection and retraction functions tend to accumulate.

To solve this second issue, a possibility is to project the current iterate onto the manifold at each step of the trust regions solver :
```diff
    # Compute the tentative next iterate (the proposal)
    retract!(p.M, o.x_proposal, o.x, o.η, o.retraction_method)
+    project!(p.M, o.x_proposal,o.x_proposal)
```
But again, I do not know if this is sensible.

Here is the code mentionned above, where the iterates leave the manifold and projecting `o.δ` onto the tangent space does not help: (The last line computes the distance of the last iterate to the manifold; if the iterate belongs to the manifold, the distance is almost zero.)
```julia
using Printf, Manifolds, ManifoldsBase, Manopt

n = 100
p = 2
M = Oblique(p,n)

# Cost matrix
C = randn(n,n) ;
C = C * C' ;

Vinit = project(M,randn(p,n))

function cost(M,V)
    # V -> Tr(C*V'*V)
    sum(vec((V*C) .* V))
end

function riem_grad(M,V)
    project(M,V,2*V*C')
end

function riem_hess(M,V,W)
    m = sum((V*C').*V,dims=1)
    WS = W * C - W .* m
    2. * project(M,V,WS)
end

Vrec = trust_regions(M,cost,riem_grad,riem_hess,Vinit;
                        stopping_criterion=StopAfterIteration(5000))

@printf("Distance to the manifold : %.2e\n",norm(Vrec-project(M,Vrec)))
```
P.S. I might have made a mistake in the Hessian function. As to the preconditioner, however, I think it is simply the identity, so it should not be a problem here.

**Comment by kellertuer** [comment]
In my experience (though from the sphere, but oblique is a power of the sphere) is that if you plug a non-tangent vector into the exponential map (we do not do retractions in the setting you use) the result might leave the manifold.

So if something in your computations leads to a tangent vector not being tangent, exp might be wrong.

The `ProjectionRetraction`(I hope we defined it for Oblique) should be another easy fix.
(Without testing, since I am just closing off for the day) add a `retraction=ProjectionRetraction()` to the trust regions call. This should be more stable since the `ProjectionRetraction` is not exactly the exp you use now, but numerically more stable (and locally “fine enough”)

A remark on the new example – can you maybe fix the seed? For me it might be random whether that fails or not otherwise (depending on to what my seed is initialised / set).

For the fix: Sure instead of (or maybe additionally to) projecting the tangent vector one can also project each iterate (or proposal). Here I already have a concrete example where that leads to problems. There is no projection onto `SymmetricPositiveDefinite(n)`, so that has to definetly be an option to activate / deactivate.

Before introducing these, I still would like to find the time (or maybe you find it) so see _where_ this error first appears – i.e. which line of code makes a tangent vector non-tangent (I think this might introduce the second error you report) or otherwise (less likely) introduce the first non-manifold-point (without the reason being a non-tangent-vector).  Since if there is a bug in the program, the projections you propose would just fix a symptom of the bug but not the bug itself.

If it is not a bug but a numerical instability – sure we should introduce either of the projections or even both as features that can be activated.

**Comment by IWalds** [comment]
Regarding the seed, I used 10 :
`Random.seed!(10)`

Here is a plot of the distance of the "tangent" vectors to the tangent space across the first 50 trust regions iterations :
![dist_eta_and_grad](https://user-images.githubusercontent.com/97890775/149904217-ddfb64a1-7bb8-4bbb-9225-e03395896fca.png)
The blue line represents the distance to the tangent of `o.η`, computed right before the instruction
`    retract!(p.M, o.x_proposal, o.x, o.η, o.retraction_method)`
inside the trust-regions solver. The red line is the distance to the tangent of the gradient, recorded each time my custom gradient function is called.

According to this plot, there is not one specific point at which the vector leaves the manifold. It is rather a slow numerical drift. (This is different from my first example, where the instability seems to be in the truncated conjugate gradient solver, and appears suddenly at iteration 17 of the trust regions solver.)

Calling the solver with the option `retraction=ProjectionRetraction()`, as you suggest, solves this issue. For my current needs, a good solution thus seems to be:
1. project onto the tangent, to avoid instabilities in the truncated conjugate gradient solver;
2. use a projection retraction instead of the exponential one in the trust regions solver.

**Comment by kellertuer** [comment]
Thanks for the analysis and the nice plot. I am not sure whether your gradient is a little of or the eta is affected by rounding errors (hopefully not by a bug). I am also not sure where the drift should come from.

Just to check – you currently do both the projection in the truncated CG as well as ProjectionRetraction? I hope the ProjectionRetraction alone should work well.

**Comment by IWalds** [comment]
Yes, I both project in the truncated CG and use ProjectionRetraction. I tried using ProjectionRetraction only, but it does not solve the instability in the truncated conjugate gradient solver. Indeed, I found an example (not one of the two above) where, at some iteration of the trust regions solver, the `o.η` generated by conjugate gradient is far from the tangent space (and thus, very different from what it should be in theory), seemingly because of an instability similar to the one I described in my first message. This does not prevent the output of the trust regions solver from belonging to the manifold, since using ProjectionRetraction forces the iterates to stay on the manifold, but it is still not the expected behavior.

**Comment by kellertuer** [comment]
Then we should definetly take a look why this is happening, i.e. whether this is
a) an instability in the method itself
b) a bug in your functions (e.g. the gradient) or
c) a bug in the current code of Manopt.jl

I am not yet sure when I find time the next few days to spin up your example using the `ValidationManifold` checking where we actually leave the manifold – maybe – if around – @TomRiem might take a look if he is interested and has time. He wrote most of this solver.

**Comment by kellertuer** [comment]
I discussed with Tom yesterday and we also looked at the Matlab implementation, where in

https://github.com/NicolasBoumal/manopt/blob/6c55f0e9ebc784199dd103bc67d619090d0e87be/manopt/solvers/trustregions/tCG.m#L279-L287

the `o.δ` (Julia variable name) is projected to the tangent space. So let's introduce that as an option to tCG (and hence also to TR).
If you see how to do it, feel free to do a PR – I keep it on my todo list and hope to do that soon otherwise.
